From 1880441e86b2f6ca2d352a0715b3709044f8088f Mon Sep 17 00:00:00 2001
From: Ammar Faizi <ammarfaizi2@gmail.com>
Date: Sun, 28 Nov 2021 19:07:57 +0700
Subject: [PATCH 2/2] chrome/renderer: Create simple extension with
 `thread->registerExtension`

This is my first attempt to create a preload Javascript under chromium.
It works fine with headful mode, but it's probably won't work when we
use the chromium as headless mode.

How to test this patch:
 1) Apply this patch.
 2) Compile.
 3) Run the chrome binary.
 4) Open console.
 5) Write `js_preload()`.

Signed-off-by: Ammar Faizi <ammarfaizi2@gmail.com>
---
 chrome/renderer/chrome_content_renderer_client.cc | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/chrome/renderer/chrome_content_renderer_client.cc b/chrome/renderer/chrome_content_renderer_client.cc
index 2bc63616d6..01e1b32fe8 100644
--- a/chrome/renderer/chrome_content_renderer_client.cc
+++ b/chrome/renderer/chrome_content_renderer_client.cc
@@ -378,6 +378,15 @@ ChromeContentRendererClient::ChromeContentRendererClient()
 
 ChromeContentRendererClient::~ChromeContentRendererClient() {}
 
+class JSPreloadExtension : public v8::Extension {
+ public:
+  JSPreloadExtension(const char* extension_name,
+                     const char* javascript_code)
+      : v8::Extension(extension_name, javascript_code)
+  {
+  }
+};
+
 void ChromeContentRendererClient::RenderThreadStarted() {
   RenderThread* thread = RenderThread::Get();
 
@@ -433,6 +442,12 @@ void ChromeContentRendererClient::RenderThreadStarted() {
   thread->AddObserver(chrome_observer_.get());
   thread->AddObserver(subresource_filter_ruleset_dealer_.get());
 
+  thread->RegisterExtension(
+    std::make_unique<JSPreloadExtension>(
+      "v8/JSPreload",
+      "var js_preload = function () { alert(\"Hello World!\"); }"
+    )
+  );
   thread->RegisterExtension(extensions_v8::LoadTimesExtension::Get());
 
   base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
-- 
2.30.2

